src/new_build_api.mli

```diff
@@ -0,0 +1,196 @@
+    script is then evuluated lazily in order to produce the requested
+    targets.
+
+    Build scripts are dynamic and can adapt to external changes such
+    as files and directories being modified, however it is encouraged
+    to write as much as possible using static constructions, as
+    detailed later. The reason for that is that when executing a static
+    script, Dune can partially see the future and what effects the
+    script is going to have on the file system. This allows Dune to
+    guide the build in order to go as fast as possible.
+
+    Dune makes the assumption that build scripts are pure. In
+    particular, it is not allowed to use global variables or access the
+    file system directly. Instead, one must use the primitives declared
+    in this module.
+*)
+
+(** The [t] data type represent the type of build scripts. ['a] is the
+    type of values that are provided to the script at execution time
+    and ['b] is the type of values it produces. *)
+type ('a, 'b) t
+
+(** [arr f] is the simplest form of build scripts. [f] is an OCaml
+    function that will be executed when the build script is
+    evaluated.
+
+    Warning: [f] must not access global variables or files.
+*)
+val arr : ('a -> 'b) -> ('a, 'b) t
+
+(** [return] is the same as [arr (fun x -> x)]. *)
+val return : ('a, 'a) t
+
+(** [const v] is a node that always evaluate to the same value. *)
+val const : 'a -> (unit, 'a) t
+
+module O : sig
+  (** The following operators allow to compose [t] values in order to
+      build complex build scripts. *)
+
+  val ( >>> ) : ('a, 'b) t -> ('b, 'c) t -> ('a, 'c) t
+  val ( ^>> ) : ('a -> 'b) -> ('b, 'c) t -> ('a, 'c) t
+  val ( >>^ ) : ('a, 'b) t -> ('b -> 'c) -> ('a, 'c) t
+
+  val ( <<< ) : ('b, 'c) t -> ('a, 'b) t -> ('a, 'c) t
+  val ( ^<< ) : ('b -> 'c) -> ('a, 'b) t -> ('a, 'c) t
+  val ( <<^ ) : ('b, 'c) t -> ('a -> 'b) -> ('a, 'c) t
+
+  val ( *** ) : ('a, 'b) t -> ('c, 'd) t -> ('a * 'c, 'b * 'd) t
+  val ( &&& ) : ('a, 'b) t -> ('a, 'c) t -> ('a, 'b * 'c) t
+end
+
+module Ext : sig
+  type ('a, 'b) build = ('a, 'b) t
+
+  (** An [Ext.t] value represent an external action. For instance a
+      command to run. *)
+  type t =
+    { deps   : Path.Set.t  (** List of files that we'll be read when
+                               execution [acton]. *)
+    ; locks  : Path.t list (** List of lock files that must be held
+                               while [action] is running. *)
+    ; action : Action.t    (** The actual action to execute. *)
+    }
+
+  (** [add_dep path] is the same as:
+
+      {[
+        arr (fun ext ->
+          { ext with deps = Path.Set.add path deps })
+      ]}
+
+      However, because [path] is given statically, Dune is able to
+      know earlier about this dependency and can parallelize the build
+      better.
+
+      Note that [add_dep] can only be called inside a call to [ext].
+      I.e. the following code is valid:
```

Comment:
it's just that it might not be clear what call to `Build.ext` this dependency is associated to. Or maybe it should just be the next one. In fact, once the `Ext.t` is evaluated, we can check that the files we though are dependencies are actually dependencies, so that shouldn't be an issue.

For instance this is a case where dune can't realize that `path` is associated only to the first `Build.ext`. But we could just say that this code is silly and reject it with a runtime check.

```ocaml
Build.add_dep path
>>>
Build.fanout
   (Build.ext ~targets:[...] Build.return)
   (Build.arr ignore
    >>>
    Build.ext ~targets:[...] (...))
```